File: src-tauri/src/model/zoom.rs
use serde::{Serialize, Deserialize};

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
pub struct ZoomTarget {
    pub t_ms: u64,
    pub x: f64,
    pub y: f64,
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
pub struct ZoomSegment {
    pub start_ms: u64,
    pub end_ms: u64,
    pub ease_in_ms: u64,
    pub ease_out_ms: u64,
    pub scale: f64,
    pub targets: Vec<ZoomTarget>,
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
pub struct ZoomModel {
    pub version: u32,
    pub segments: Vec<ZoomSegment>,
}

/// Cubic smoothstep with input clamped to [0,1].
pub fn smoothstep(p: f64) -> f64 {
    let p = p.clamp(0.0, 1.0);
    p * p * (3.0 - 2.0 * p)
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct ZoomAt {
    pub scale: f64,
    pub cx: f64,
    pub cy: f64,
}

fn target_at(seg: &ZoomSegment, t_ms: u64) -> (f64, f64) {
    let ts = &seg.targets;
    if ts.len() == 1 {
        return (ts[0].x, ts[0].y);
    }
    let first = ts[0].t_ms;
    let last = ts[ts.len() - 1].t_ms;
    let t = t_ms.clamp(first, last);
    for w in ts.windows(2) {
        let (a, b) = (&w[0], &w[1]);
        if t >= a.t_ms && t <= b.t_ms {
            let span = (b.t_ms - a.t_ms).max(1) as f64;
            let f = (t - a.t_ms) as f64 / span;
            return (a.x + (b.x - a.x) * f, a.y + (b.y - a.y) * f);
        }
    }
    (ts[ts.len() - 1].x, ts[ts.len() - 1].y)
}

pub fn zoom_at(model: &ZoomModel, t_ms: u64) -> ZoomAt {
    for seg in &model.segments {
        if t_ms < seg.start_ms || t_ms >= seg.end_ms {
            continue;
        }
        let rel = (t_ms - seg.start_ms) as f64;
        let dur = (seg.end_ms - seg.start_ms) as f64;
        let ein = seg.ease_in_ms as f64;
        let eout = seg.ease_out_ms as f64;
        let e_in = if ein > 0.0 {
            smoothstep(rel / ein)
        } else {
            1.0
        };
        let e_out = if eout > 0.0 {
            smoothstep((dur - rel) / eout)
        } else {
            1.0
        };
        let e = if rel < ein && rel > dur - eout {
            e_in.min(e_out) // overlapping ramps: take the smaller
        } else if rel < ein {
            e_in
        } else if rel > dur - eout {
            e_out
        } else {
            1.0
        };
        let scale_t = 1.0 + (seg.scale - 1.0) * e;
        let (tx, ty) = target_at(seg, t_ms);
        let m = 0.5 / scale_t;
        let cx = tx.clamp(m, 1.0 - m);
        let cy = ty.clamp(m, 1.0 - m);
        return ZoomAt {
            scale: scale_t,
            cx,
            cy,
        };
    }
    ZoomAt {
        scale: 1.0,
        cx: 0.5,
        cy: 0.5,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn smoothstep_endpoints_and_mid() {
        assert!((smoothstep(0.0) - 0.0).abs() < 1e-9);
        assert!((smoothstep(1.0) - 1.0).abs() < 1e-9);
        assert!((smoothstep(0.5) - 0.5).abs() < 1e-9);
        assert!((smoothstep(-3.0) - 0.0).abs() < 1e-9); // clamp
        assert!((smoothstep(7.0) - 1.0).abs() < 1e-9);  // clamp
    }

    #[test]
    fn model_round_trips() {
        let m = ZoomModel {
            version: 1,
            segments: vec![ZoomSegment {
                start_ms: 0,
                end_ms: 2000,
                ease_in_ms: 300,
                ease_out_ms: 400,
                scale: 2.0,
                targets: vec![ZoomTarget {
                    t_ms: 0,
                    x: 0.25,
                    y: 0.75,
                }],
            }],
        };
        let j = serde_json::to_string(&m).unwrap();
        let back: ZoomModel = serde_json::from_str(&j).unwrap();
        assert_eq!(m, back);
    }

    fn fixture() -> ZoomModel {
        ZoomModel {
            version: 1,
            segments: vec![ZoomSegment {
                start_ms: 0,
                end_ms: 2000,
                ease_in_ms: 300,
                ease_out_ms: 400,
                scale: 2.0,
                targets: vec![ZoomTarget {
                    t_ms: 0,
                    x: 0.25,
                    y: 0.75,
                }],
            }],
        }
    }

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 1e-6
    }

    #[test]
    fn zoom_at_outside_is_identity() {
        let z = zoom_at(&fixture(), 2500);
        assert!(close(z.scale, 1.0) && close(z.cx, 0.5) && close(z.cy, 0.5));
    }

    #[test]
    fn zoom_at_ease_in_mid() {
        let z = zoom_at(&fixture(), 150); // smoothstep(0.5)=0.5 -> scale 1.5
        assert!(close(z.scale, 1.5), "{}", z.scale);
        assert!(close(z.cx, 1.0 / 3.0), "{}", z.cx); // clamp(0.25, 0.333..,0.666..)
        assert!(close(z.cy, 2.0 / 3.0), "{}", z.cy); // clamp(0.75, ...)
    }

    #[test]
    fn zoom_at_plateau_full_scale() {
        let z = zoom_at(&fixture(), 1000);
        assert!(close(z.scale, 2.0));
        assert!(close(z.cx, 0.25) && close(z.cy, 0.75)); // m=0.25
    }

    #[test]
    fn zoom_at_ease_out_mid() {
        let z = zoom_at(&fixture(), 1800); // (2000-1800)/400=0.5 -> 0.5 -> scale 1.5
        assert!(close(z.scale, 1.5));
    }
}

File: src/lib/zoom.ts
import type { WebcamOverlay } from "./webcam";

export interface ZoomTarget { t_ms: number; x: number; y: number }
export interface ZoomSegment {
  start_ms: number; end_ms: number; ease_in_ms: number; ease_out_ms: number;
  scale: number; targets: ZoomTarget[];
}
export interface ZoomModel { version: number; segments: ZoomSegment[]; webcam?: WebcamOverlay }
export interface ZoomAt { scale: number; cx: number; cy: number }

const clamp = (v: number, lo: number, hi: number) => Math.min(hi, Math.max(lo, v));

export function smoothstep(p: number): number {
  const c = clamp(p, 0, 1);
  return c * c * (3 - 2 * c);
}

function targetAt(seg: ZoomSegment, t: number): [number, number] {
  const ts = seg.targets;
  if (ts.length === 1) return [ts[0].x, ts[0].y];
  const first = ts[0].t_ms;
  const last = ts[ts.length - 1].t_ms;
  const tc = clamp(t, first, last);
  for (let i = 0; i < ts.length - 1; i++) {
    const a = ts[i], b = ts[i + 1];
    if (tc >= a.t_ms && tc <= b.t_ms) {
      const span = Math.max(1, b.t_ms - a.t_ms);
      const f = (tc - a.t_ms) / span;
      return [a.x + (b.x - a.x) * f, a.y + (b.y - a.y) * f];
    }
  }
  return [ts[ts.length - 1].x, ts[ts.length - 1].y];
}

export function zoomAt(model: ZoomModel, tMs: number): ZoomAt {
  for (const seg of model.segments) {
    if (tMs < seg.start_ms || tMs >= seg.end_ms) continue;
    const rel = tMs - seg.start_ms;
    const dur = seg.end_ms - seg.start_ms;
    const ein = seg.ease_in_ms, eout = seg.ease_out_ms;
    const eIn = ein > 0 ? smoothstep(rel / ein) : 1;
    const eOut = eout > 0 ? smoothstep((dur - rel) / eout) : 1;
    let e: number;
    if (rel < ein && rel > dur - eout) e = Math.min(eIn, eOut);
    else if (rel < ein) e = eIn;
    else if (rel > dur - eout) e = eOut;
    else e = 1;
    const scale = 1 + (seg.scale - 1) * e;
    const [tx, ty] = targetAt(seg, tMs);
    const m = 0.5 / scale;
    return { scale, cx: clamp(tx, m, 1 - m), cy: clamp(ty, m, 1 - m) };
  }
  return { scale: 1, cx: 0.5, cy: 0.5 };
}
